src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java

```diff
@@ -0,0 +1,230 @@
+  }
+
+  /**
+   * Constructs a ContainerRegistryAuthSupplier using the specified credentials.
+   *
+   * @see Builder
+   */
+  public static Builder forEcrClient(final AmazonECR ecr) {
+    return new Builder(ecr);
+  }
+
+  /**
+   * A Builder of ContainerRegistryAuthSupplier.
+   * <p>
+   * The default value for the minimum expiry time of an access token is one minute. When the
+   * ContainerRegistryAuthSupplier is asked for a RegistryAuth, it will check if the existing
+   * AccessToken for the AWS AuthorizationData expires within this amount of time. If it does, then
+   * the AuthorizationData is refreshed before being returned.
+   * </p>
+   */
+  public static class Builder {
+
+    private final AmazonECR ecr;
+    private long minimumExpiryMillis = TimeUnit.MINUTES.toMillis(1);
+
+    public Builder(final AmazonECR ecr) {
+      this.ecr = ecr;
+    }
+
+    /**
+     * Changes the minimum expiry time used to refresh AccessTokens before they expire. The default
+     * value is one minute.
+     */
+    public Builder withMinimumExpiry(long duration, TimeUnit timeUnit) {
+      this.minimumExpiryMillis = TimeUnit.MILLISECONDS.convert(duration, timeUnit);
+
+      return this;
+    }
+
+    public ContainerRegistryAuthSupplier build() {
+      final Clock clock = Clock.SYSTEM;
+
+      return new ContainerRegistryAuthSupplier(ecr, clock, minimumExpiryMillis,
+          new EcrCredentials(ecr));
+    }
+  }
+
+  private final AmazonECR ecr;
+
+  // TODO (mbrown): change to java.time.Clock once on Java 8
+  private final Clock clock;
+  private EcrCredentials credentials;
+  private final long minimumExpiryMillis;
+
+  @VisibleForTesting
+  ContainerRegistryAuthSupplier(final AmazonECR ecr, final Clock clock,
+                                final long minimumExpiryMillis, final EcrCredentials credentials) {
+    Preconditions.checkArgument(ecr != null, "ecr");
+    this.ecr = ecr;
+    this.clock = clock;
+    this.minimumExpiryMillis = minimumExpiryMillis;
+    this.credentials = credentials;
+  }
+
+  /**
+   * Get an accessToken to use, possibly refreshing the token if it expires within the
+   * minimumExpiryMillis.
+   */
+  private AuthorizationData getAccessToken() throws IOException {
+    // synchronize attempts to refresh the accessToken
+    synchronized (ecr) {
+      if (needsRefresh(credentials.getAuthorizationData())) {
+        credentials.refresh();
+      }
+    }
+
+    Preconditions.checkState(credentials.getAuthorizationData() != null,
+        "authorizationData should have been refreshed");
+
+    return credentials.getAuthorizationData();
+  }
+
+  boolean needsRefresh(final AuthorizationData accessToken) {
+    if (accessToken == null) {
+      // has not yet been fetched
+      return true;
+    }
+
+    final Date expirationTime = accessToken.getExpiresAt();
+
+    // Don't refresh if expiration time hasn't been provided.
+    if (expirationTime == null) {
+      return true;
```

Comment:
Do we need to test this? Will `accessToken.getExpiresAt()` ever be null?

Reply:
It's technically nullable - https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-ecr/src/main/java/com/amazonaws/services/ecr/model/AuthorizationData.java